kubernetes_env/benchmark/benchmark.py

```diff
@@ -20,115 +24,137 @@
+                percentages.append(datum["Percent"])
+            df = pd.DataFrame({
+                "Latency (ms)": latency,
+                "Percent": percentages
+            })
             dfs.append(df)
-    final_df = pd.concat(dfs)
-    fig, ax = plt.subplots(figsize=(12, 6))
-    sns.lineplot(data=final_df,
-                 x="Response time (ms)",
-                 y="Percentiles",
-                 hue="Filter",
-                 ci=False,
-                 marker='o')
-    plt.title(title)
-    timestamp = time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime(time.time()));
-    plot_name = "-".join(names) + timestamp
+    return dfs, title
+
+
+def plot(dfs, filters, title, fortio=True):
+    timestamp = time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime(time.time()))
+    plot_name = "-".join(filters) + timestamp
+    if fortio:
+        for df in dfs:
+            sns.lineplot(data=df, x="Latency (ms)", y="Percent")
+        plt.legend(labels=filters)
+        plt.title(f"Fortio: {title}")
+    else:
+        plot = sns.displot(dfs, kind="ecdf")
+        plot.set(title=plot_name)
     util.check_dir(GRAPHS_DIR)
     plt.savefig(f"{GRAPHS_DIR}/{plot_name}.png")
     log.info("Finished plotting. Check out the graphs directory!")
     return util.EXIT_SUCCESS
 
 
-def run_fortio(platform, threads, qps, run_time, file_name):
+def run_fortio(url, platform, threads, qps, run_time, file_name):
     util.check_dir(DATA_DIR)
-    _, _, gateway_url = kube_env.get_gateway_info(platform)
     output_file = str(DATA_DIR.joinpath(f"{file_name}.json"))
     fortio_dir = str(FORTIO_DIR)
     cmd = f"{fortio_dir} "
     cmd += f"load -c {threads} -qps {qps} -jitter -t {run_time}s -json {output_file} "
-    cmd += f"http://{gateway_url}/productpage"
+    cmd += f"{url}"
     with open(output_file, "w") as f:
         fortio_res = util.exec_process(cmd,
                                        stdout=subprocess.PIPE,
                                        stderr=subprocess.PIPE)
         return fortio_res
 
 
-def start_benchmark(fortio, filter_dirs, platform, threads, qps, time):
+def burst_loop(url, threads, qps, run_time, queue):
+    def send_request(_):
+        try:
+            # What should timeout be?
+            res = requests.get(url)
+            ms = res.elapsed.total_seconds() * 1000
+            return ms
+        except requests.exceptions.ReadTimeout:
+            pass
+
+    log.info("Starting burst...")
+
+    with ThreadPoolExecutor(max_workers=threads) as p:
+        current = datetime.now()
+        end = current + timedelta(seconds=run_time)
+        while current < end:
+            results = list(
+                p.map(send_request, range(qps), chunksize=qps // threads))
+            current += timedelta(seconds=1)
+            queue.put(results)
+
+
+def do_burst(url, platform, threads, qps, run_time):
+    queue = Queue()
+    _, _, gateway_url = kube_env.get_gateway_info(platform)
+    p = Process(target=burst_loop, args=(
```

Comment:
You are launching a process but you immediately wait for it again. It looks like you do not even need this and can call the function directly.